Panics when unmarshaling from underlying type to derived type

```
panic: reflect.Set: value of type string is not assignable to type RuleType [recovered]
	panic: reflect.Set: value of type string is not assignable to type RuleType
```

--- a/marshal.go
+++ b/marshal.go
@@ -313,20 +313,22 @@ func unmarshalAttribute(node *Resource, value reflect.Value, tokens []string) er
 
 	errs := make([]error, 0)
 	attrValue := reflect.ValueOf(attr)
+	vtype := value.Type()
 	if attrValue.Kind() == reflect.Map || attrValue.Kind() == reflect.Slice {
 		// the attribute was a JSON object or array,
 		// so leverage JSON marshaling/unmarshaling.
 		data, err := json.Marshal(attr)
 		errs = append(errs, err)
 
-		vtype := value.Type()
 		ptr := newItem(vtype)
 		ptrValue := ptr.Interface()
 		errs = append(errs, json.Unmarshal(data, &ptrValue))
 
 		setValue(vtype, value, ptr)
-	} else {
+	} else if attrValue.Type().AssignableTo(vtype) {
 		value.Set(reflect.ValueOf(attr))
+	} else if attrValue.CanConvert(vtype) {
+		value.Set(attrValue.Convert(vtype))
 	}
 
 	return errors.Join(errs...)

--- a/marshal_test.go
+++ b/marshal_test.go
@@ -8,10 +8,13 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+type typedef string
+
 type SimpleItem struct {
-	ID     string `jsonapi:"primary,items"`
-	Value1 string `jsonapi:"attr,value1"`
-	Value2 string `jsonapi:"attr,value2,omitempty"`
+	ID     string  `jsonapi:"primary,items"`
+	Value1 string  `jsonapi:"attr,value1"`
+	Value2 string  `jsonapi:"attr,value2,omitempty"`
+	Value3 typedef `jsonapi:"attr,value3,omitempty"`
 	Omit   any
 }
 
@@ -403,6 +406,7 @@ func TestUnmarshalResource(t *testing.T) {
 			Type: "items",
 			Attributes: map[string]any{
 				"value1": "foo",
+				"value3": "bar",
 			},
 			Links: jsonapi.Links{},
 			Meta:  jsonapi.Meta{},
@@ -411,7 +415,7 @@ func TestUnmarshalResource(t *testing.T) {
 		err := jsonapi.UnmarshalResource(&in, &out)
 		assert.NoError(t, err)
 
-		want := SimpleItem{ID: "1", Value1: "foo"}
+		want := SimpleItem{ID: "1", Value1: "foo", Value3: typedef("bar")}
 		assert.EqualValues(t, want, out)
 	})
 

